### PR TITLE
AC2 Ocelots Class - Catherine Bandarchuk

### PR DESCRIPTION
**Pull request by CatherineBandarchuk**

### CODE REVIEW

File: tests/test_wave_01.py [hunk 1]
```diff
@@ -118,13 +111,10 @@ def test_moves_movie_from_watchlist_to_empty_watched():
     # Assert
     assert len(updated_data["watchlist"]) == 0
     assert len(updated_data["watched"]) == 1
+    assert updated_data["watched"][0]["title"] == MOVIE_TITLE_1
+    assert updated_data["watched"][0]["genre"] == GENRE_1
+    assert updated_data["watched"][0]["rating"] == RATING_1
```
Comment (kelsey-steven-ada) [inline tests/test_wave_01.py:116]: Nice checks for all the movie's data!

File: tests/test_wave_01.py [hunk 2]
```diff
@@ -142,13 +132,10 @@ def test_moves_movie_from_watchlist_to_watched():
     # Assert
     assert len(updated_data["watchlist"]) == 1
     assert len(updated_data["watched"]) == 2
+    assert updated_data["watched"][0]["title"] == movie_to_watch["title"]
+    assert updated_data["watched"][0]["genre"] == movie_to_watch["genre"]
+    assert updated_data["watched"][0]["rating"] == movie_to_watch["rating"]
```
Comment (kelsey-steven-ada) [inline tests/test_wave_01.py]: The README isn't explicit about what order we should add movies to the list `watched`. If we want our tests to be independent from a specific implementation that adds movies to a particular location in the `watched` list, we could use the `in` keyword to check if the variable `movie_to_watch` is in the watched list instead of looking at a specific index:

```python
assert len(updated_data["watchlist"]) == 1
assert len(updated_data["watched"]) == 2
assert movie_to_watch in updated_data["watched"]
```

File: tests/test_wave_03.py [hunk 3]
```diff
@@ -43,24 +43,23 @@ def test_friends_unique_movies():
...
     # Assert
     assert len(friends_unique_movies) == 3
-
-    raise Exception("Test needs to be completed.")
-    # *************************************************************************************************
-    # ****** Add assertions here to test that the correct movies are in friends_unique_movies **********
-    # **************************************************************************************************
-
-@pytest.mark.skip()
+    assert len(set(list_of_title)) == len(list_of_title)
```
Comment (kelsey-steven-ada) [inline tests/test_wave_03.py]: Here, we compare the length of the set of titles vs the length of the list of titles. This confirms that the number of titles is correct, but it doesn't guarantee that their values are what we expect. I think we could also say this check duplicates the info that the previous assert `assert len(friends_unique_movies) == 3` gives us.

If we explore the `clean_wave_3_data` test data in the `test_constants` file, we could reason about which titles should be in our result and directly assert on those. This would also reduce the time the test takes to run since we could avoid creating a list of the titles and then a set from that list.
```python
assert len(friends_unique_movies) == 3
assert INTRIGUE_3 in friends_unique_movies
assert HORROR_1 in friends_unique_movies
assert FANTASY_4 in friends_unique_movies
```

File: viewing_party/party.py [hunk 4]
```diff
@@ -1,23 +1,146 @@
 # ------------- WAVE 1 --------------------
+import statistics
 
 def create_movie(title, genre, rating):
-    pass
+    new_movie = dict()
```
Comment (kelsey-steven-ada) [inline viewing_party/party.py:7]: I'd suggest moving the `new_movie` declaration under the invalid data check so that we don't allocate space for `new_movie` unless we know it will be used.

File: viewing_party/party.py [hunk 5]
```diff
@@ -1,23 +1,146 @@
...
 def create_movie(title, genre, rating):
-    pass
+    new_movie = dict()
+    if title == None or genre == None or rating == None:
+        return None
+    else:
+        new_movie["title"] = title
+        new_movie["genre"] = genre
+        new_movie["rating"] = rating
+        return new_movie
```
Comment (kelsey-steven-ada) [inline viewing_party/party.py]: If we do not explicitly need an `else` clause, I suggest leaving it off, as it is more more statement that needs to be evaluated, and often causes the main flow of our function that we want to be very clear to readers to be indented. We could remove the `else` here without changing the outcome of our function:

```python
if title == None or genre == None or rating == None:
    return None

new_movie = dict()
new_movie["title"] = title
new_movie["genre"] = genre
new_movie["rating"] = rating
return new_movie

# another option could be:
if title == None or genre == None or rating == None:
    return None

return {
    "title": title,
    "genre": genre,
    "rating": rating 
}
```

File: viewing_party/party.py [hunk 6]
```diff
@@ -1,23 +1,146 @@
...
 
+def add_to_watched(user_data, movie):
+    updated_user_data = {
+        "watched": []
+    }
+    updated_user_data["watched"].insert(0,movie)
+    return updated_user_data
```
Comment (kelsey-steven-ada) [inline viewing_party/party.py]: Something we need to be really careful about is avoiding overwriting data that we want to preserve. In the description for the `add_to_watched` function, the README says:

> the value of user_data will be a dictionary with a key "watched", and a value which is a list of dictionaries representing the movies the user has watched

This means that the function parameter `user_data` is already a dictionary that contains a `watched` list, and which might have values inside. Line 15 create a brand new data structure `updated_user_data` that gets returned. This ignores the `user_data` parameter and we lose any data that was in `user_data`.

How could we re-write this function to preserve any data we get in the input `user_data`?

File: viewing_party/party.py [hunk 7]
```diff
@@ -1,23 +1,146 @@
...
+
+def add_to_watchlist(user_data, movie):
+    updated_user_data = {
+        "watchlist": []
+    }
```
Comment (kelsey-steven-ada) [inline viewing_party/party.py]: The comment in `add_to_watched` about preserving data from the input `user_data` applies here as well. How would we re-write this function to ensure we don't lose info from the `user_data` parameter?

File: viewing_party/party.py [hunk 8]
```diff
@@ -1,23 +1,146 @@
...
+        new_movie["title"] = title
+        new_movie["genre"] = genre
+        new_movie["rating"] = rating
+        return new_movie
 
+def add_to_watched(user_data, movie):
+    updated_user_data = {
+        "watched": []
+    }
+    updated_user_data["watched"].insert(0,movie)
```
Comment (kelsey-steven-ada) [inline viewing_party/party.py]: Nice use of `insert`! To help readers visually process a line easier and quicker, I would recommend adding a space between the parameters in the `insert` function call: `insert(0, movie)`

File: viewing_party/party.py [hunk 9]
```diff
@@ -1,23 +1,146 @@
...
+def add_to_watchlist(user_data, movie):
+    updated_user_data = {
+        "watchlist": []
+    }
+    updated_user_data["watchlist"].insert(0,movie)
+    return updated_user_data
+
+def watch_movie(user_data, movie_title):
+    if movie_title == None or user_data["watchlist"] == None:
+        return user_data
```
Comment (kelsey-steven-ada) [inline viewing_party/party.py]: Nice handling for returning `None` for missing data! In Python, when we test if a variable is pointing to `None` the PEP8 style guide recommends using `is` or `is not` for comparison rather than operators like `==`. 

`None` is a special object in Python, it’s a `singleton`, meaning there is only ever one instance of `None` in memory. All variables holding `None` are actually pointing to the same instance. By using `is` to compare against `None`, we avoid issues that could be caused by a class’s custom implementation of the comparison operators, and we gain some speed over a check using `==` since we are checking against identity rather than looking up if the variable's held type has a comparison function that should be used.

Something else to consider: here we're explicitly checking if the values are None, but what if we got an empty string for `movie_title` or an empty list for `user_data["watchlist"]`? If we wanted to check for those at the same time as we check for None, we could rely on the truthy/falsy values of the parameters and write something like: 
```python
if not movie_title or not user_data["watchlist"]:
    return user_data
``` 

If you’d like more info on this, here’s a couple resources to get started =]
https://peps.python.org/pep-0008/#programming-recommendations
http://jaredgrubb.blogspot.com/2009/04/python-is-none-vs-none.html

File: viewing_party/party.py [hunk 10]
```diff
@@ -1,23 +1,146 @@
...
+    else:
```
Comment (kelsey-steven-ada) [inline viewing_party/party.py]: I would recommend removing the `else` here since it doesn't change the outcome of our code.

File: viewing_party/party.py [hunk 11]
```diff
@@ -1,23 +1,146 @@
...
+
+def watch_movie(user_data, movie_title):
+    if movie_title == None or user_data["watchlist"] == None:
+        return user_data
+    else: 
+        for movie in user_data["watchlist"]:
+            if movie["title"] == movie_title:
+                user_data["watched"].insert(0,movie)
+                user_data["watchlist"].remove(movie)
+    return user_data
```
Comment (kelsey-steven-ada) [inline viewing_party/party.py]: This function works, but as an aside it contains an "anti-pattern" that is important to be aware of.  An element of the list that is being iterated over is being removed inside the for loop.  This won't cause problems in this particular case, but consider this slightly different implementation:
```
def watch_movie(user_data, title):
    for i in range(len(user_data["watchlist"])):
        item = user_data["watchlist"][i]

        if item["title"] == title:
            user_data["watched"].append(item)
            user_data["watchlist"].remove(item)
    return user_data
```
If we were to test this in a scenario where `user_data["watchlist"]` has 2 elements, and the first is the movie that we are moving to `user_data["watched"]`, the second iteration through the loop will throw an index error, because it will try to access the 2nd element in `user_data["watchlist"]`, but there no longer is a second element.  

As a coding pattern, this is one that I recommend avoiding because it has so much potential for unforeseen errors. One fix is to return immediately after making the swap (ie, move line 36 inside the if-block so that the loop ends as soon as the swap is made).  Another fix is to use the loop to find the item, but not make the swap until after the loop ends.

File: viewing_party/party.py [hunk 12]
```diff
@@ -1,23 +1,146 @@
...
 # -----------------------------------------
 # ------------- WAVE 2 --------------------
 # -----------------------------------------
+def get_watched_avg_rating(user_data):
+    avg_rating = 0.0
+    sum_ratings = 0
+    if len(user_data["watched"]) == 0:
+        return 0.0
+    else:
```
Comment (kelsey-steven-ada) [inline viewing_party/party.py]: Looking at the pattern of using `else` blocks across the file, I might suggest for the next couple assignments to practice explicitly not writing `else` at first then adding it in later if it's necessary.

File: viewing_party/party.py [hunk 13]
```diff
@@ -1,23 +1,146 @@
...
+        for element in user_data["watched"]:
+            sum_ratings += element["rating"]
+        avg_rating = sum_ratings/len(user_data["watched"])
+        return avg_rating
 
-
+def get_most_watched_genre(user_data):
+    most_watched_genre = ""
+    list_of_genre = list()
```
Comment (kelsey-steven-ada) [inline viewing_party/party.py]: I'm seeing a pattern of declaring variables at the top of a function before validation checks. If you have a checklist of things to look for when reviewing your code before submission, it could be helpful to add an item around checking if variables are being declared before we know if they will be used.

File: viewing_party/party.py [hunk 14]
```diff
@@ -1,23 +1,146 @@
...
+    if len(user_data["watched"]) == 0:
+        return None
```
Comment (kelsey-steven-ada) [inline viewing_party/party.py:45]: The more pythonic way to check if our variables have content is to use the truthy/falsy value which lets us check for an empty list or None all in one:
```python
if not user_data["watched"]:
    return None
```

File: viewing_party/party.py [hunk 15]
```diff
@@ -1,23 +1,146 @@
...
+    else:
+        for i in range(len(user_data["watched"])):
+            list_of_genre.append(user_data["watched"][i]["genre"])
+        most_watched_genre = statistics.mode(list_of_genre)
+        return most_watched_genre
```
Comment (kelsey-steven-ada) [inline viewing_party/party.py]: Great use of `mode` to get the most watched genre! 

This would have about the same performance, but another way we could go about this could be to create a dictionary that holds a frequency map of the genres (where they keys are the genres and the values are their counts), then return the genre with the max value:

```python
genre_freqs = {}
for movie in user_data["watched"]:
    genre = movie['genre']
    if genre not in genre_freqs:
        genre_freqs[genre] = 1
    else:
        genre_freqs[genre] += 1

max_value = max(genre_freqs, key=genre_freqs.get)
return max_value
```

File: viewing_party/party.py [hunk 16]
```diff
@@ -1,23 +1,146 @@
...
+def get_watched_avg_rating(user_data):
+    avg_rating = 0.0
+    sum_ratings = 0
+    if len(user_data["watched"]) == 0:
+        return 0.0
+    else:
+        for element in user_data["watched"]:
+            sum_ratings += element["rating"]
+        avg_rating = sum_ratings/len(user_data["watched"])
+        return avg_rating
```
Comment (kelsey-steven-ada) [inline viewing_party/party.py]: Nice approach! Another way we could do this is using list comprehensions and the `sum` function:

```python
# The list comprehension inside the `sum` call pulls the value of the 
# `rating` key out of each item in the list `user_data['watched']`
total = sum(movie['rating'] for movie in user_data['watched'])
average_rating = total / len(user_data['watched'])
return average_rating
```

File: viewing_party/party.py [hunk 17]
```diff
@@ -1,23 +1,146 @@
...
+    friends_movie_list = list()
+    unique_movies = list()
+    if len(user_data["watched"]) == 0:
+        unique_movies = []
+        return unique_movies
+    else:
+        for i in range(len(user_data["friends"])):
+            for j in range(len(user_data["friends"][i]["watched"])):
+                friends_movie_list.append(user_data["friends"][i]["watched"][j]["title"])
+        friends_movie_set = set(friends_movie_list)
```
Comment (kelsey-steven-ada) [inline viewing_party/party.py]: These lines get a little long and hard to easily understand because of the explicit indexing and nesting of the `user_data` structure. In general, I recommend using for-in loops for their ease of reading over ranged for-loops unless we need an index to do something special that is more than accessing a single value.

In this solution, we only use the indices of the ranged for loops to access values for a single list (we aren't trying to index into the same location of multiple lists). This means we could simplify our syntax and make the code a bit easier to read by using `for in` loops with descriptive variable names. We could also use a little less memory by removing `friends_movie_list` and adding friends' movie titles directly to `friends_movie_set`:

```python
friends_movie_set = set()
for friend in user_data["friends"]:
    for movie in friend["watched"]:
        friends_movie_set.add(movie["title"])
```

Comment (kelsey-steven-ada) [inline viewing_party/party.py]: Love the use of a set here!

File: viewing_party/party.py [hunk 18]
```diff
@@ -1,23 +1,146 @@
...
+        for k in range(len(user_data["watched"])):
+            if user_data["watched"][k]["title"] not in friends_movie_set:
+                unique_movies.append(user_data["watched"][k])
+        return unique_movies
```
Comment (kelsey-steven-ada) [inline viewing_party/party.py]: Similar to the feedback above, we could simplify the loop syntax for filling `unique_movies` by using a for-in loop over a ranged loop. What could that look like?

File: viewing_party/party.py [hunk 19]
```diff
@@ -1,23 +1,146 @@
...
+    
+def get_friends_unique_watched(user_data):
+    friends_movie_list = list()
+    friends_unique_movies = list()
+    if len(user_data["watched"]) == 0:
+        for i in range(len(user_data["friends"])):
+            for j in range(len(user_data["friends"][i]["watched"])):
+                friends_movie_list.append(user_data["friends"][i]["watched"][j])
```
Comment (kelsey-steven-ada) [inline viewing_party/party.py]: Most of the code in the `if` block is duplicated in the `else` as well. If we need to change something about how we get the list of friends' movies we would need to make changes in multiple places. When we see code duplication like this, we can usually pull it out of the if/else and leave only the code that differs in the if/else.

```python
for i in range(len(user_data["friends"])):
    for j in range(len(user_data["friends"][i]["watched"])):
        friends_movie_list.append(user_data["friends"][i]["watched"][j])

if len(user_data["watched"]) == 0:
    return friends_unique_movies

else:
    for k in range(len(friends_movie_list)):
        if (friends_movie_list[k] not in user_data["watched"]) and ((friends_movie_list[k] not in friends_unique_movies)):
            friends_unique_movies.append(friends_movie_list[k])

    return friends_unique_movies
```

Comment (kelsey-steven-ada) [inline viewing_party/party.py]: How could we reimagine this code using for-in loops rather than `range`?

File: viewing_party/party.py [hunk 20]
```diff
@@ -1,23 +1,146 @@
...
+        return friends_unique_movies
+    else:
+        for i in range(len(user_data["friends"])):
+            for j in range(len(user_data["friends"][i]["watched"])):
+                friends_movie_list.append(user_data["friends"][i]["watched"][j])
+        for k in range(len(friends_movie_list)):
+            if (friends_movie_list[k] not in user_data["watched"]) and ((friends_movie_list[k] not in friends_unique_movies)):
```
Comment (kelsey-steven-ada) [inline viewing_party/party.py]: This line is pretty long, I suggest breaking this statement apart. The PEP 8 style guide recommends limiting lines to a max of 79 characters to make code easier to read, especially with multiple editor windows open: https://peps.python.org/pep-0008/#maximum-line-length. 

Outside of the style guide, when you're in situations where other people will be reading and working in the same code, the ability to quickly and easily comprehend what the code is doing is often more valuable than reducing line count.

One option could be to create variables for the each side of the `and` operator:
```python
is_movie_not_in_watched = friends_movie_list[k] not in user_data["watched"]
is_movie_not_in_unique_list = friends_movie_list[k] not in friends_unique_movies
if is_movie_not_in_watched and is_movie_not_in_unique_list:
```

File: viewing_party/party.py [hunk 21]
```diff
@@ -1,23 +1,146 @@
...
+                friends_unique_movies.append(friends_movie_list[k])
+        return friends_unique_movies#_set
         
 # -----------------------------------------
 # ------------- WAVE 4 --------------------
 # -----------------------------------------
-
+def get_available_recs(user_data):
+    list_of_recommended_movies = list()
+    unique_friends_movie = get_friends_unique_watched(user_data)
```
Comment (kelsey-steven-ada) [inline viewing_party/party.py:82]: Great code reuse!

File: viewing_party/party.py [hunk 22]
```diff
@@ -1,23 +1,146 @@
...
+    if not user_data["watched"] or not unique_friends_movie or not user_data["subscriptions"]:
+        return list_of_recommended_movies
+    else:
+        for i in range(len(unique_friends_movie)):
+            if unique_friends_movie[i]["host"] in user_data["subscriptions"]:
+                list_of_recommended_movies.append(unique_friends_movie[i])
+        if not list_of_recommended_movies:
+            return list_of_recommended_movies
+        else:
+            return list_of_recommended_movies
```
Comment (kelsey-steven-ada) [inline viewing_party/party.py]: Because the `if` and `else` contain the same code, we could remove those and just keep the line `return list_of_recommended_movies`

File: viewing_party/party.py [hunk 23]
```diff
@@ -1,23 +1,146 @@
...
         
 # -----------------------------------------
 # ------------- WAVE 4 --------------------
 # -----------------------------------------
-
+def get_available_recs(user_data):
+    list_of_recommended_movies = list()
+    unique_friends_movie = get_friends_unique_watched(user_data)
+    if not user_data["watched"] or not unique_friends_movie or not user_data["subscriptions"]:
+        return list_of_recommended_movies
```
Comment (kelsey-steven-ada) [inline viewing_party/party.py]: Nice choice to exit early if the relevant data is missing.

File: viewing_party/party.py [hunk 24]
```diff
@@ -1,23 +1,146 @@
...
 # -----------------------------------------
+def get_new_rec_by_genre(user_data):
+    recommended_movies = list()
+    list_of_genre = list()
+    if not user_data["watched"]:
+        return recommended_movies
+    else:
+        for movie in user_data["watched"]:
+            list_of_genre.append(movie["genre"])
+        most_ofen_genre = statistics.mode(list_of_genre)
```
Comment (kelsey-steven-ada) [inline viewing_party/party.py]: It looks like lines 122-124 duplicates what we do in `get_most_watched_genre`. How could we re-write this to use our previous function `get_most_watched_genre` instead?

File: viewing_party/party.py [hunk 25]
```diff
@@ -1,23 +1,146 @@
...
+        list_of_movies = get_friends_unique_watched(user_data)
+        if not list_of_movies or not list_of_genre:
```
Comment (kelsey-steven-ada) [inline viewing_party/party.py]: We could forgo this if-check without impacting the outcome of the function. If `list_of_movies` is empty, then we'd never enter the loop on line 129, we don't need extra checks to prevent it. 
``` python
list_of_movies = get_friends_unique_watched(user_data)

for i in range(len(list_of_movies)):
    if list_of_movies[i]["genre"] == most_ofen_genre:
        recommended_movies.append(list_of_movies[i])

return recommended_movies
```

File: viewing_party/party.py [hunk 26]
```diff
@@ -1,23 +1,146 @@
...
+def get_rec_from_favorites(user_data):
+    recommended_movies = list()
+    favorite_movies = list()
+    if not user_data["watched"]:
+        return recommended_movies
+    else:
+        unique_movies = get_unique_watched(user_data) #user's watched movies, different from friend's movies
+        for i in range(len(user_data["favorites"])):
+            favorite_movies.append(user_data["favorites"][i])
+        for movie in unique_movies:
```
Comment (kelsey-steven-ada) [inline viewing_party/party.py]: Great use for a for-in loop!

File: viewing_party/party.py [hunk 27]
```diff
@@ -1,23 +1,146 @@
...
 
+def get_rec_from_favorites(user_data):
+    recommended_movies = list()
+    favorite_movies = list()
+    if not user_data["watched"]:
+        return recommended_movies
+    else:
+        unique_movies = get_unique_watched(user_data) #user's watched movies, different from friend's movies
+        for i in range(len(user_data["favorites"])):
+            favorite_movies.append(user_data["favorites"][i])
```
Comment (kelsey-steven-ada) [inline viewing_party/party.py]: It looks like `favorite_movies` is an unchanged copy of `user_data["favorites"]`. If that's the case, then we could factor out `favorite_movies` and use `user_data["favorites"]` directly on line 144 below:

```python
for movie in unique_movies:
    if movie in user_data["favorites"]:
        recommended_movies.append(movie) 
```

**Review by kelsey-steven-ada — COMMENTED** [review]
Congrats on completing your first project at Ada! I’ve added some suggestions & questions, let me know if there's anything I can clarify ^_^

File: viewing_party/party.py [hunk 28]
```diff
@@ -1,23 +1,116 @@
...
+
+def watch_movie(user_data, title):
+    if not title or not user_data["watchlist"]:
+        return user_data
+    for i in range(len(user_data["watchlist"])):
+        movie = user_data["watchlist"][i]
+        if movie["title"] == title:
+            user_data["watched"].append(movie)
+            user_data["watchlist"].remove(movie)
+    return user_data
```
Comment (kelsey-steven-ada) [inline viewing_party/party.py:29]: We're still running into an antipattern of continuing to loop after removing an element of the list we're looping over. I would recommend keeping the for-in loop syntax and moving the return inside the if-statement something like:
```python
for movie in user_data["watchlist"]:
    if movie["title"] == title:
        user_data["watched"].append(movie)
        user_data["watchlist"].remove(movie)
        return user_data
```

**Review by kelsey-steven-ada — COMMENTED** [review]
I like how you incorporated the feedback! There was only one note around the antipattern of continuing to loop after making changes to a data structure that I think is really important to make sure we understand what is happening and why we want to avoid it. 

Otherwise, around code-styles, in the future I'd consider adding some blank lines inside of functions to visually separate chunks of code. For example, I like to add a blank line after my validation check(s) to give readers a visual separation between the validation code and the main flow of the function. I might also use a blank line to separate chunks of code that perform subtasks of the overall goal.

**Comment by CatherineBandarchuk** [comment]
Thank you for your feedback! :)
Sometimes I’m overdoing testing and thought maybe several movies with same title will be held in user_data :)) but anyway it’s definitely the bug if I remove element and keep looping with index i(and the length of the list already less by one)…
I’m just curious, the function len() does get invoke every time in for loop if I’m using index in list?

Thanks again,
Kate

Sent from my iPhone

On Nov 7, 2022, at 10:08 AM, Kelsey Steven ***@***.***> wrote:

﻿

@kelsey-steven-ada commented on this pull request.

I like how you incorporated the feedback! There was only one note around the antipattern of continuing to loop after making changes to a data structure that I think is really important to make sure we understand what is happening and why we want to avoid it.

Otherwise, around code-styles, in the future I'd consider adding some blank lines inside of functions to visually separate chunks of code. For example, I like to add a blank line after my validation check(s) to give readers a visual separation between the validation code and the main flow of the function. I might also use a blank line to separate chunks of code that perform subtasks of the overall goal.

________________________________

In viewing_party/party.py<https://github.com/ada-ac2/viewing-party/pull/1#discussion_r1015735171>:

> +    for i in range(len(user_data["watchlist"])):
+        movie = user_data["watchlist"][i]
+        if movie["title"] == title:
+            user_data["watched"].append(movie)
+            user_data["watchlist"].remove(movie)
+    return user_data


We're still running into an antipattern of continuing to loop after removing an element of the list we're looping over. I would recommend keeping the for-in loop syntax and moving the return inside the if-statement something like:

for movie in user_data["watchlist"]:
    if movie["title"] == title:
        user_data["watched"].append(movie)
        user_data["watchlist"].remove(movie)
        return user_data

**Comment by kelsey-steven-ada** [comment]
That is a great question. If we declare our loop with `for i in range(len(<list_name>)):` the length gets evaluated when the loop is created, but it does not keep getting updated as the loop progresses. This is part of why we see an IndexError after the first couple iterations if we try to run code like:
```python
list_a = ["a", "b", "c", "d"]
for i in range(len(list_a)):
    print(list_a[i])
    del list_a[i]
```

The code above will output "a", then "c", then crash with `IndexError: list index out of range`
- Our range is initialized to a sequence of values from 0 to 3
- In the first iteration, `i` is 0 so we print then delete "a"
- In the second iteration, `i` is 1. We deleted "a", so now "b" is at index 0 and "c" is at index 1. This means we will skip "b" and print then delete "c" at index 1
- In our third iteration, `i` is 2 and "a" and "c" have been deleted leaving "b" at index 0 and "d" at index 1. When we hit the line `print(list_a[i])` we will now crash since `i` is 2, but we have no element at index 2.

**Comment by CatherineBandarchuk** [comment]
Thank you! It is really helpful.

Sent from my iPhone

On Nov 8, 2022, at 9:37 AM, Kelsey Steven ***@***.***> wrote:

﻿

That is a great question. If we declare our loop with for i in range(len(<list_name>)): the length gets evaluated when the loop is created, but it does not keep getting updated as the loop progresses. This is part of why we see an IndexError after the first couple iterations if we try to run code like:

list_a = ["a", "b", "c", "d"]
for i in range(len(list_a)):
    print(list_a[i])
    del list_a[i]

The code above will output "a", then "c", then crash with IndexError: list index out of range

  *   Our range is initialized to a sequence of values from 0 to 3
  *   In the first iteration, i is 0 so we print then delete "a"
  *   In the second iteration, i is 1. We deleted "a", so now "b" is at index 0 and "c" is at index 1. This means we will skip "b" and print then delete "c" at index 1
  *   In our third iteration, i is 2 and "a" and "c" have been deleted leaving "b" at index 0 and "d" at index 1. When we hit the line print(list_a[i]) we will now crash since i is 2, but we have no element at index 2.
